src/main/java/me/kavin/piped/server/handlers/auth/UserHandlers.java

```diff
@@ -109,10 +132,277 @@ public static byte[] loginResponse(String user, String pass)
         }
     }
 
-    public static byte[] deleteUserResponse(String session, String pass) throws IOException {
+    public static HttpResponse oidcLoginRequest(OidcProvider provider, String redirectUri) throws Exception {
+        if (StringUtils.isBlank(redirectUri)) {
+            return HttpResponse.ofCode(400).withHtml("redirect is a required parameter");
+        }
+
+        URI callback = new URI(Constants.PUBLIC_URL + "/oidc/" + provider.name + "/callback");
+        CodeVerifier codeVerifier = new CodeVerifier();
+        OidcData data = new OidcData(redirectUri, codeVerifier);
+        String state = data.getState();
+
+        DatabaseHelper.setOidcData(data);
+
+        AuthenticationRequest oidcRequest = new AuthenticationRequest.Builder(
+                new ResponseType("code"),
+                new Scope("openid"),
+                provider.clientID, callback
+        )
+                .endpointURI(provider.authUri)
+                .codeChallenge(codeVerifier, CodeChallengeMethod.S256)
+                .state(new State(state))
+                .nonce(data.getOidNonce()).build();
+
+        if (redirectUri.equals(Constants.FRONTEND_URL + "/login")) {
+            return HttpResponse.redirect302(oidcRequest.toURI().toString());
+        }
+        return HttpResponse.ok200().withHtml(
+                "<!DOCTYPE html><html style=\"color-scheme: dark light;\"><body>" +
+                        "<h3>Warning:</h3> You are trying to give <pre style=\"font-size: 1.2rem;\">" +
+                        redirectUri +
+                        "</pre> access to your Piped account. If you wish to continue click " +
+                        "<a style=\"text-decoration: underline;color: inherit;\"href=\"" +
+                        oidcRequest.toURI().toString() +
+                        "\">here</a></body></html>");
+    }
+
+    public static HttpResponse oidcLoginCallback(OidcProvider provider, URI requestUri) throws Exception {
+        AuthenticationSuccessResponse authResponse = parseOidcUri(requestUri);
+
+        OidcData data = DatabaseHelper.getOidcData(authResponse.getState().toString());
+        if (data == null) {
+            return HttpResponse.ofCode(400).withHtml(
```

Comment:
Since we don't provide any styling or `color-scheme: dark light;`, `withPlainText` would probably be better to not become flashed in dark mode